File: 30.包含min函数的栈/30.包含min函数的栈.py
# -*- coding:utf-8 -*-
class Solution:
    def __init__(self):
        self.stack = []
        self.minstack = []
    def push(self, node):
        # write code here
        self.stack.append(node)
        if len(self.minstack)==0:
            self.minstack.append(node)
        elif node <= self.min():
            self.minstack.append(node)
    def pop(self):
        # write code here
        node = self.stack.pop()
        if node == self.min():
            self.minstack.pop()
    def top(self):
        # write code here
        return self.stack[-1]
    def min(self):
        # write code here
        return self.minstack[-1]

File: 30.包含min函数的栈/30.包含min函数的栈.cpp
class Solution {
    vector<int> stack;
    vector<int> minstack;

public:
    void push(int value) {
        stack.push_back(value);
        if (minstack.size()==0)
        	minstack.push_back(value);
        else if (value <= min())
        {
            minstack.push_back(value);
        }
        		
    }
    void pop() {
        int val = top();
        stack.erase(stack.end()-1);    
        if (val==min())
        	minstack.erase(minstack.end()-1);
    }
    int top() {
        return stack[stack.size()-1];
    }
    int min() {
        return minstack[minstack.size()-1];
    }
};